File: #227_Basic_Calculator_II/Solution2.java
import java.util.*;

/* Use stack to save previous result 
 * Much more elegent
 */
public class Solution2 {
    public int calculate(String s) {
    	if (s == null || s.length() == 0) return 0;
    	Deque<Long> stack = new LinkedList();
    	char sign = '+';
    	long num = 0;
    	for (int i = 0; i < s.length(); i++) {
    		if (Character.isDigit(s.charAt(i))) {
    			num = num * 10 + s.charAt(i) - '0';
    		}
    		if ((!Character.isDigit(s.charAt(i)) && s.charAt(i) != ' ') || i == s.length() - 1) {
    			switch (sign) {
    				case '+':
    					stack.push(num);
    					break;
    				case '-':
    					stack.push(-num);
    					break;
    				case '*':
    					stack.push(stack.pop() * num);
    					break;
    				case '/':
    					stack.push(stack.pop() / num);
    					break;
    			}
    			sign = s.charAt(i);
    			num = 0;
    		}
    	}
    	long result = 0;
    	while (!stack.isEmpty()) {
    		result += stack.pop();
    	}
    	return (int) result;
    }

    public static void main(String[] args) {
    	Solution2 sol = new Solution2();
		System.out.println(sol.calculate("14/3*2"));
    }
}

File: #224_Basic_Calculator/Solution2.py
class Solution(object):
	def calculate(self, s):
		"""
		:type s: str
		:rtype: int
		"""
		if not s:
			return 0
		stack = []
		re = 0
		num = 0
		sign = 1 # 1 for +, -1 for -
		for c in s:
			if c.isdigit():
				num = num * 10 + ord(c) - ord('0')
			elif c == '+':
				re += num * sign
				num = 0
				sign = 1
			elif c == '-':
				re += num * sign
				num = 0
				sign = -1
			elif c == '(':
				stack.append(re)
				stack.append(sign)
				re = 0
				num = 0
				sign = 1
			elif c == ')':
				re += num * sign
				re *= stack.pop()
				re += stack.pop()
				num = 0
				sign = 1
		return re + num * sign

sol = Solution()
print sol.calculate("(1+(4+5+2)-3)+(6+8)")